File: src/strategies/ReportStrategies.js
/**
 * Strategy — один и тот же отчёт считается разными алгоритмами.
 * Контекст (ReportContext) не знает, какой именно алгоритм вызван.
 */

class ReportStrategy {
  execute(_transactions, _categories) {
    throw new Error("Стратегия должна реализовать execute()");
  }
}

class ByCategoryStrategy extends ReportStrategy {
  execute(transactions, categories) {
    const map = new Map();

    for (const tx of transactions) {
      const current = map.get(tx.categoryId) || 0;
      map.set(tx.categoryId, current + tx.amount);
    }

    return [...map.entries()].map(([categoryId, total]) => {
      const category = categories.find((item) => item.id === categoryId);
      return {
        key: categoryId,
        label: category ? category.name : "Без категории",
        type: category ? category.type : "expense",
        total,
      };
    });
  }
}

class ByDayStrategy extends ReportStrategy {
  execute(transactions) {
    const map = new Map();

    for (const tx of transactions) {
      const day = tx.date.slice(0, 10);
      const current = map.get(day) || { income: 0, expense: 0 };
      current[tx.type] += tx.amount;
      map.set(day, current);
    }

    return [...map.entries()]
      .sort((a, b) => a[0].localeCompare(b[0]))
      .map(([day, totals]) => ({
        key: day,
        label: formatDay(day),
        income: totals.income,
        expense: totals.expense,
        total: totals.income - totals.expense,
      }));
  }
}

class ByTypeStrategy extends ReportStrategy {
  execute(transactions) {
    const income = transactions
      .filter((tx) => tx.type === "income")
      .reduce((sum, tx) => sum + tx.amount, 0);
    const expense = transactions
      .filter((tx) => tx.type === "expense")
      .reduce((sum, tx) => sum + tx.amount, 0);

    return [
      { key: "income", label: "Доходы", total: income },
      { key: "expense", label: "Расходы", total: expense },
    ];
  }
}

class ReportContext {
  constructor(strategy) {
    this.strategy = strategy;
  }

  setStrategy(strategy) {
    this.strategy = strategy;
  }

  generate(transactions, categories) {
    return this.strategy.execute(transactions, categories);
  }
}

function createReportStrategy(name) {
  switch (name) {
    case "byDay":
      return new ByDayStrategy();
    case "byType":
      return new ByTypeStrategy();
    case "byCategory":
    default:
      return new ByCategoryStrategy();
  }
}

function formatDay(isoDate) {
  const [year, month, day] = isoDate.split("-");
  return `${day}.${month}.${year}`;
}

module.exports = {
  ReportContext,
  createReportStrategy,
  ByCategoryStrategy,
  ByDayStrategy,
  ByTypeStrategy,
};

File: app/src/main/java/com/finmate/app/domain/strategy/ReportStrategies.kt
package com.finmate.app.domain.strategy

import com.finmate.app.data.model.Category
import com.finmate.app.data.model.Transaction
import com.finmate.app.data.model.TransactionType

data class ReportRow(
    val key: String,
    val label: String,
    val total: Double = 0.0,
    val income: Double? = null,
    val expense: Double? = null,
    val type: TransactionType? = null
)

/** Strategy: варианты построения отчёта. */
interface ReportStrategy {
    fun execute(transactions: List<Transaction>, categories: List<Category>): List<ReportRow>
}

class ByCategoryStrategy : ReportStrategy {
    override fun execute(transactions: List<Transaction>, categories: List<Category>): List<ReportRow> {
        return transactions
            .groupBy { it.categoryId }
            .map { (categoryId, items) ->
                val category = categories.find { it.id == categoryId }
                ReportRow(
                    key = categoryId,
                    label = category?.name ?: "Без категории",
                    total = items.sumOf { it.amount },
                    type = category?.type
                )
            }
    }
}

class ByDayStrategy : ReportStrategy {
    override fun execute(transactions: List<Transaction>, categories: List<Category>): List<ReportRow> {
        return transactions
            .groupBy { it.date.take(10) }
            .toSortedMap()
            .map { (day, items) ->
                val income = items.filter { it.type == TransactionType.INCOME }.sumOf { it.amount }
                val expense = items.filter { it.type == TransactionType.EXPENSE }.sumOf { it.amount }
                ReportRow(
                    key = day,
                    label = formatDay(day),
                    income = income,
                    expense = expense,
                    total = income - expense
                )
            }
    }

    private fun formatDay(iso: String): String {
        val parts = iso.split("-")
        if (parts.size != 3) return iso
        return "${parts[2]}.${parts[1]}.${parts[0]}"
    }
}

class ByTypeStrategy : ReportStrategy {
    override fun execute(transactions: List<Transaction>, categories: List<Category>): List<ReportRow> {
        val income = transactions.filter { it.type == TransactionType.INCOME }.sumOf { it.amount }
        val expense = transactions.filter { it.type == TransactionType.EXPENSE }.sumOf { it.amount }
        return listOf(
            ReportRow(key = "income", label = "Доходы", total = income, type = TransactionType.INCOME),
            ReportRow(key = "expense", label = "Расходы", total = expense, type = TransactionType.EXPENSE)
        )
    }
}

class ReportContext(private var strategy: ReportStrategy) {
    fun setStrategy(strategy: ReportStrategy) {
        this.strategy = strategy
    }

    fun generate(transactions: List<Transaction>, categories: List<Category>): List<ReportRow> {
        return strategy.execute(transactions, categories)
    }
}

fun createReportStrategy(name: String): ReportStrategy = when (name) {
    "byDay" -> ByDayStrategy()
    "byType" -> ByTypeStrategy()
    else -> ByCategoryStrategy()
}
